### PR TITLE
feat: network sync - add main sync logic

### PR DESCRIPTION
**Pull request by Prithpal-Sooriya**
## Explanation

This adds support for the "main-sync" on network syncing feature. This will get all a users local networks, and saved remote networks; then will append/update/delete networks synced across their devices.

We will add the controller integration in a following PR.

## References

https://consensyssoftware.atlassian.net/browse/NOTIFY-1040

### `@metamask/profile-sync-controller`

- **CHANGED**: Renamed the `sync.ts` file to `sync-mutations.ts`
- **ADDED**: `utils` to the `UserStorageController` subpath.
- **ADDED**: `findNetworksToUpdate()` function which will perform the "main-sync" logic for local and remote network configurations.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  return {
+    localMap,
+    remoteMap,
+    localKeySet,
+    remoteKeySet,
+    existingRemoteKeySet,
+  };
+};
+
+type MatrixResult = 'Do Nothing' | 'Local Wins' | 'Remote Wins';
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts:53]: See [figma](https://www.figma.com/board/fZTsDekl5xbHo8IheI3MMc/Network-Syncing-Notes?node-id=0-1&node-type=canvas&t=tWpXvIVguaMgdpYG-0).

This is the main result matrix that intersecting networks are compared against.

File: packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+      ...updatedNetworks.remoteNetworksToUpdate,
+    ];
+
+    // List of new local networks
+    const newLocalNetworks = getNewLocalNetworks({
+      originalList: localNetworks,
+      missingLocalNetworks: missingNetworks.missingLocalNetworks,
+      localNetworksToRemove: updatedNetworks.localNetworksToRemove,
+      localNetworksToUpdate: updatedNetworks.localNetworksToUpdate,
+    });
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts:227]: I'm not entirely convinced we need this combined local network lists.

Depending on the integration, we may choose to swap this for the separated lists (might make it easier to determine which lists need updating or removing)

File: packages/profile-sync-controller/src/controllers/user-storage/network-syncing/types.ts [hunk 3]
```diff
@@ -1,4 +1,9 @@
-import type { NetworkConfiguration } from '@metamask/network-controller';
+import type { NetworkConfiguration as _NetworkConfiguration } from '@metamask/network-controller';
+
+// TODO - replace shim once we update NetworkController type
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/network-syncing/types.ts:3]: This is temporary until the network controller is updated.

File: packages/profile-sync-controller/src/controllers/user-storage/network-syncing/types.ts [hunk 4]
```diff
@@ -11,3 +16,19 @@ export type RemoteNetworkConfiguration = NetworkConfiguration & {
...
+    v: '1',
+  };
+};
+
+export const toNetworkConfiguration = (
+  network: RemoteNetworkConfiguration,
+): NetworkConfiguration => {
+  const { v: _v, d: _d, ...originalNetwork } = network;
+  return originalNetwork;
+};
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/network-syncing/types.ts:34]: Schema conversions.

I may rename this file into `schema.ts` in near future 🤔

File: packages/profile-sync-controller/src/controllers/user-storage/utils.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Returns the difference between 2 sets.
+ * NOTE - this is temporary until we can support Set().difference method
```
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/utils.ts:3]: Aww, the original implementation used the new JS [Set utilities](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/difference).

However both our TS version, and transpilers did not play well with it.

Still kept the Set data structure (a bit nicer to work with) and remade these utilities.

File: packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts (hunk 2 above)
Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts:230]: That is fine, we can change this in a future PR depending if the integration with the external `NetworkController` interface.

**Review by mathieuartu — APPROVED** [review]
Excellent job! I might re-use some of the way you laid out your logic, as I think it makes better sense for future use of the batch PUT endpoint.

Nice coverage as well, wooo

**Pull request merged**